%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
%\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%\def\includegraphic{}
%\def\includegraphics{}
\usepackage{graphicx}


%%% Put your definitions there:
\startlocaldefs
\endlocaldefs

% Our variable names
\newcommand{\chrOnePop}{Chr1\_Population}
\newcommand{\FullPop}{Full\_Population}
\newcommand{\FullFam}{Chr1\_Family}


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Methodology}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{Scalable Clustering of Genotype Information using MapReduce}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
   addressref={aff1,aff2},                   % id's of addresses, e.g. {aff1,aff2}
%   noteref={n1},                        % id's of article notes, if any
   email={Aidan.O'Brien@csiro.au}   % email address
]{\inits{AR}\fnm{Aidan R} \snm{O'Brien}}
\author[
   addressref={aff3, aff4},
   email={f.buske@garvan.org.au}
]{\inits{FA}\fnm{Fabian A} \snm{Buske}}
\author[
   addressref={aff1},
   corref={aff1},                       % id of corresponding address, if any
   email={Denis.Bauer@CSIRO.au}
]{\inits{DC}\fnm{Denis C} \snm{Bauer}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{%                           % unique id
  \orgname{Division of Computational Informatics, CSIRO}, % university, etc
  \street{11 Julius Av},                     %
  \postcode{2113}                                % post or zip code
  \city{Sydney},                              % city
  \cny{Australia}                                    % country
}
\address[id=aff2]{%
  \orgname{School of Biomedical Sciences and Pharmacy, Faculty of Health},
%  \street{},
  \postcode{2308}
  \city{Newcastle},
  \cny{Australia}
}
\address[id=aff3]{%
  \orgname{Cancer Epigenetics Program, Cancer Research Division, Kinghorn Cancer Centre, Garvan Institute of Medical Research},
  \street{384 Victoria St},
  \postcode{2010}
  \city{Sydney},
  \cny{Australia}
}
\address[id=aff4]{%
  \orgname{UNSW Medicine, University of New South Wales},
%  \street{},
  \postcode{2052}
  \city{Sydney},
  \cny{Australia}
}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\begin{artnotes}
%\note{Sample of title note}     % note to the article
%\note[id=n1]{Equal contributor} % note, connected to author
%\end{artnotes}

%\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}

\begin{abstract} % abstract
\parttitle{Background} Processing genomic information from whole genome sequence studies pose computational challenges due to the unprecedented data volume generated, which render transitional approaches insufficient. However, by utilising advancements in modern hardware accelerators and data processing we can provide the means for scalable solutions. We therefore aim to provide the interface between standard genomic data formats and advanced and scalable analysis libraries like Mahout. 
\parttitle{Results} We achieve an XX-fold speedup by using the scalable k-means MapReduce implementation over the equivalent analysis performed in R and YY-fold speedup for PCA, respectively, by comparable accuracy. Due to memory in the R-Implementation the comparison had to be limited to variant data from chromosome 1 only, however we ran the full dataset (aa GB, bb individuals {\`a} cc variants ) through Mahout demonstrating the scalability. 
\parttitle{Conclusions} Using modern compute paradigms is essential to scale to modern genomic research in an efficient sustainable way. 
\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}
\kwd{Hadoop}
\kwd{Mahout}
\kwd{Genotype}
\kwd{Clustering}
\kwd{MapReduce}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
% <put your article body there>

%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background}
Grouping individuals based on the genomic profile is a commonly performed tasks to identify population structure~\cite{Gao2007} or elucidate different haplotype involvement in diseases susceptibility~\cite{Laitman2013}.  Traditionally both the number of individuals and included genotypes, typically from SNP arrays, were relatively small and libraries in Bioconductor sufficient. However, recent technological advances in whole genome sequencing have made population-scale sequencing feasible. It is hence economical to generate studies with sample sizes currently reserved for larger consortia such as the 1000 genomes project~\cite{1KG2012} or TCGA~\cite{TCGA2013}. At the same time, whole genome sequencing enables the inclusion of rare or even somatic mutations in the analysis, increasing the feature space by orders of magnitude. This drastic increase in both sample numbers and features per sample requires a massively parallele approach to data processing. 

As a result of these big data challenges, MapReduce approaches are increasingly being used in bioinformatics (for reviews see~cite{Zou2013, Qiu2010,Taylor2010}). This is especially the case for sequence analysis tasks, such as read mapping~\cite{Schatz2009}, duplicate removal~\cite{Jourdren2012}, and variant calling~\cite{Langmead2009, McKenna2010} or Genome Wide Analysis Study based tasks~\cite{Huang2013, Guo2014}.

At the same time, sophisticated machine learning methodologies are increasingly adapted to utilise MapReduce paradigms overcoming callings that arise due to their iterative nature~\cite{Chu2009}. Specifically, the Mahout project\url{https://mahout.apache.org/} has been developed extensively~\cite{Ranger2007, Owen2011} and was successfully applied in the clinical informatics space~\cite{Dong2013}.

We therefore link the two areas by providing an interface between Mahout and the standard variant data format, VCF~\cite{1KG2012}, which opens up the application of Mahout's different machine learning algorithms to be applied to genotype-based tasks.   
To demonstrate the capability we cluster variant datasets from the 1000 genomes project to determine population structure as well as ancestry using different algorithms available in Mahout. In the first two sections we benchmark the performance and accuracy of Mahout's implementations against standard R based implementations on a reduced version of the data and in the last section we investigate the performance on the full dataset, demonstrating how seamless Mahout can transit to population-scale tasks.   

%SeqWare~{OConnor2010} Libraries \cite{Doering2008,Schumacher2014,Nordberg2013}







\section*{Results and discussion}
\subsection*{Runtime}
Text for this section.

\subsection*{Accuracy}
Text for this section.

\subsection*{Real World Example}
Text for this section.

\section*{Conclusion}
Note,  Domain-specific language for linear algebraic operations on top of Apache Spark \url{http://spark.apache.org/}
ADAM~\cite{Massie2013}


\section*{Methods}
\subsection*{Datasets}
We used variant call format (VCF) files from the 1000 Genomes Project.
These VCF files contain information about the genetic variants between each of the 1092 individuals and a reference genome derived from GRCh37. 
Each of the VCF files are partitioned by chromosome.
Metadata available from the 1000 Genomes Consortium specifies additional attributes for each individual, including population and familial data.
To be able to compare with R, we created a subset of the entire human variant dataset by using only variants from chromosome 1 (\chrOnePop) where we investigate the correct grouping of the population structure. 
The population structure is also investigated using the full dataset, i.e. all chromosomes, using Mahout libraries (\FullPop). 
In addition, we also investigate whether the full dataset allows the accurate prediction of the more detailed family association (\FullFam).
Table~\ref{datasets} contains an overview of the datasets.

\subsection*{Converting to arrays}
Mahout requires sequence files as input to its clustering algorithms and provides libraries for writing vectors to these files.
These libraries require a feature vector and optionally a sample name. The vectors must contain an integer for each element.
Therefore, we need to preprocess the data from the VCF files into a suitable format. The variant data values, which are tab separated,
must be converted to an integer. For this, we simply determine the Hamming distance between each tab separated value, and the reference genome.
Therefore, a value identical to the reference will therefore be 0, a heterozygous variant will be 1 and a homozygous variant will be 2.


\subsection*{Mahout K-means clustering}
We Cluster samples from a VCF file containing the variants from chromosome one. The content of the VCF file represents

Mahout provides libraries for clustering data, 'sequences', although the data must first be converted from the VCF format to
sequence files. The sequence files should consist of key-value pairs, where
Using a custom Java program, we transpose the



\subsection*{Clustering quality}
We scored the clusters with the Adjusted Rand index. 
%TODO: can you provide the formular
This index compares two different clusterings and assigns a value based on their similarity. The value is zero for independent clusterings and one for identical clusterings. We compared the Mahout k-means
clusters and the R k-means clusters to the known clusters based on meta data from the 1000 genome project. For the known clusters,
individuals were partitioned by their super population code (ASN, EUR, AFR, AMR). 
%TODO: "also on \chrOnePop" on what dataset did you compare the R performance otherwise ?
We also compared the Mahout clusters to the
R clusters for the Chr1 subset data.

\subsection*{Speed}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Availability of supporting data}
\section*{List of abbreviations}


\section*{Competing interests}
  The authors declare that they have no competing interests.

\section*{Author's contributions}
    Text for this section \ldots

\section*{Acknowledgements}
  Text for this section \ldots
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file
\bibliography{genotypeClustering}      % Bibliography file (usually '*.bib' )

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}
  \begin{figure}[h!]
  \includegraphics[type=pdf,ext=.pdf,read=.pdf, scale=0.40]{images/signature}
  \caption{\csentence{Illustration of Mahout-based clustering of genotypes.}
      The image shows how the here introduced interface converts the vcf file to a mahout usable vector-based data type. Based on k-means it illustrates how the Map-step procedure groups of vectors by their nearest cluster center and the Reduce-step averages the vectors in each group to find the new value of the updated cluster center. The mahout-produced output is then converted into a visualisation.}
      \end{figure}

\begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      Figure legend text.}
      \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
\section*{Tables}
\begin{table}[h!]
\caption{Datasets used in this study.}
      \begin{tabular}{lcc}
        \hline
           & individuals  & variants  \\ \hline
        \chrOnePop & 1024 & 3,007,196 \\
        \FullPop\ &  & \\
        \FullFam\ & 1024 & ...\\
      \end{tabular}
      \label{datasets}
\end{table}

\begin{table}[h!]
\caption{Adjusted Rand index (accuracy).}
      \begin{tabular}{lccc}
        \hline
           & Chr1 subset & 1000genomes trios  & hapmap \\ \hline
        Mahout, k=4 & 0.664 & ..&..\\
        Mahout, k=10 & 0.714 & ..&..\\
        Mahout, k=20 & 0.769 & .. & ..\\ 
        Mahout, k=30 & 0.728 & .. & ..\\ 
        R, k=4 & .. & - & -\\
        R, k=10 & .. & - & -\\
        R, k=20 & .. & - & -\\ 
        R, k=30 & .. & - & -\\ \hline
      \end{tabular}
      \label{datasetsAcc}
\end{table}

\begin{table}[h!]
\caption{Time taken to cluster samples (k=20).}
      \begin{tabular}{lccc}
        \hline
           & Chr1 subset & 1000genomes trios  & hapmap \\ \hline
        Mahout, 1 node & .. & ..&..\\
        Mahout, 2 nodes & .. & ..&..\\
        Mahout, 3 nodes & .. & .. & ..\\ 
        Mahout, 4 nodes & .. & .. & ..\\ 
        R & .. & - & -\\ \hline
      \end{tabular}
      \label{datasetsTime}
\end{table}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{backmatter}
\end{document}
