% This is LLNCS.DOC the documentation file of
% the LaTeX2e class from Springer-Verlag
% for Lecture Notes in Computer Science, version 2.4
\documentclass{llncs}
\usepackage{llncsdoc}
%

% Variable definitions
\newcommand{\variantSpark}{{\sc VariantSpark}}
\newcommand{\kMeans}{\textit{k}-means}
\newcommand{\ARI}{adjusted Rand index}



\begin{document}

\newcounter{save}\setcounter{save}{\value{section}}
{\def\addtocontents#1#2{}%
\def\addcontentsline#1#2#3{}%
\def\markboth#1#2{}%
%
\title{Using VariantSpark to uncover the link between obesity and cancer}

\author{Aidan R. O'Brien\inst{1,2}, Jason Ross\inst{1}, Robert Dunne\inst{1}, Firoz Anwar\inst{1}, \and Denis C. Bauer\inst{1}}

\institute{
CSIRO, 11 Julius Av, 2113, Sydney, Australia \\
\and
School of Biomedical Sciences and Pharmacy, Faculty of Health, University of Newcastle, 2308, Newcastle, Australia \\
}

\maketitle
%
\begin{abstract}
This paragraph shall summarize the contents of the paper
in short terms.
\end{abstract}
%
\section{Introduction}
%

Obesity is the second leading risk factor for cancer~\cite{DePergola2013}. 
For colorectal cancer (CRC) there is a 2\% increase in risk per 1 kg/m$^2$, which is similar to the pancreatic, ovarian and prostate cancer risk (about RR 1.06 - 1.10 per 5 kg/m2), while other rarer cancers such as Kidney, gallbladder and endometrial cancers have an even higher risk of up to RR 1.5 per 5 kg/m2. 
The currently observed shift in the global population from a median BMI of 25 to 30 means we must expect a 10\% to 50\% increase in cancer cases in the future, which is akin to the temporal relationship observed with cigarette smoking and lung cancer. 
While the associations between obesity and cancer are robust, almost nothing is known about the mechanism.

BMI itself has a genetic component as demonstrated in a recent meta study, which finds 97 BMI-associated loci using linear regression against BMI adjusting for study-specific covariates~\cite{Locke2015}. 
It is therefore critical to study the impact of obesity on cancer on a large well characterised cohort such as that of the cancer genome atlas (TCGA)~\cite{1KG2012}. 
Hakimi {\it et al.} investigated implication of obesity in the renal cell carcinoma mortality using TCGA data of somatic mutation, copy number, global methylation, and gene/mRNA expression~\cite{DePergola2013}. 
They used a logistic regression models to infer the associations between BMI and disease and a multivariable competing risks regression models to estimate associations between BMI and cancer specific mortality. 
They concluded that BMI is not an independent prognostic factor for cancer specific mortality after controlling for stage and grade. 
But as Locke {\it et al.}~\cite{Locke2015} pointed out in their study, only five of the 97 loci have an independent association signal, the rest seem jointly predict the metabolic phenotype.

In fact, several studies have shown that single gene or single SNP approaches often fail to identify strong feature association or find marginal effect of the feature on the target phenotype~\cite{Bureau2005,Yoo2012,Qi2012}. 
%Such approaches undermine the inter-relationship between multiple trait loci and often specify one locus a time (Pociot, Karlsen et al. 2004). In addition to that, 
The phenotypic relationship between obesity and cancer may hence manifest itself in small additive multi-loci effects, known as epistasis~\cite{Mackay2014}, which is not interrogated with logistic regression models.
Using conventional machine learning analysis to elucidate this additive effects is challenging as there are not enough samples to robustly fit the model over such a high-dimensional feature space~\cite{Chen2012}. 
As a result machine learning methods such as random forest (RF) have become popular as the underlying decision trees are build from a subsets of the features, which make the approach robust against overfitting issues~\cite{Breiman2001}. 
However, training a RF on such an expansive dataset as the TCGA is computationally challenging and traditional massively parallel approaches to data processing are not easily applicable to machine learning tasks, as they iteratively refine models based on information from the full dataset. 

We therefore utilise the {\sc Spark} engine in our recently developed framework, \variantSpark~\cite{OBrien}, for applying machine learning methods to standard variant files in Variant Call Format (VCF). 
Unlike {\sc Hadoop}, {\sc Spark} allows a more flexible software design to utilise node-parallelisation and offers in memory cashing, both critically components for efficient "big learning" tasks. 

To demonstrate \variantSpark's capability, we train a RF on the genomic profiles and BMI values of cancer patients from the TCGA dataset to detect differences in the mutational signature of tumours arising in a lean and obese background. 
In the first section we determine the cross-validation accuracy of the model and discuss overfitting potential. 
In section two record the runtime and memory consumption of the approach. 
In the last section we compare the genomic profiles of samples that can be fitted by a model very well as oppose to those that do not and discuss the influence of cancer type and tissue of origin. 




%The major outcomes for us are:
%1) Can we detect a difference in the mutational signature of tumours arising in a lean and obese background?
%2) Can we characterise the observed differences and highlight new possible modes of therapy?
%3) Can we develop a biomarker(s) that identifies this group of tumours with an obese-specific mutational signature?





%
\section{Results}
%
\subsubsection{Random forest accurately predicts BMI pan-cancer}


\begin{table}
\caption{Cross-validated results, mean accuracy (and standard error).}
\begin{center}
\renewcommand{\arraystretch}{1.4}
\setlength\tabcolsep{3pt}
\begin{tabular}{lcccc}
\hline\noalign{\smallskip}
method  & test folds & training folds & hold out \\
5-fold  & X (Y) & X (Y) & X (Y) \\
\noalign{\smallskip}
\hline
\end{tabular}
\end{center}
\end{table}

\subsubsection{Runtime}

\begin{table}
\caption{Runtime and memory consuption}
\begin{center}
\renewcommand{\arraystretch}{1.4}
\setlength\tabcolsep{3pt}
\begin{tabular}{lcccc}
\hline\noalign{\smallskip}
method  & nodes & runtime & memory \\
\noalign{\smallskip}
\hline
\end{tabular}
\end{center}
\end{table}




\subsubsection{Biological insights}


% 
\begin{figure}
\vspace{2.5cm}
\caption{This is the caption of the figure displaying a white eagle and
a white horse on a snow field}
\end{figure}


\bibliographystyle{plain}
\bibliography{recomb}  

\end{document}
